api/types/backend/backend.go

```diff
@@ -29,9 +30,31 @@ type ContainerAttachConfig struct {
 // for users of the backend to to pass it a logging configuration.
 type ContainerLogsConfig struct {
 	types.ContainerLogsOptions
-	OutStream io.Writer
+	Messages chan *LogMessage
 }
 
+// LogMessage is datastructure that represents piece of output produced by some
+// container.  The Line member is a slice of an array whose contents can be
+// changed after a log driver's Log() method returns.
+// changes to this struct need to be reflect in the reset method in
+// daemon/logger/logger.go
+type LogMessage struct {
+	Line      []byte
+	Source    string
+	Timestamp time.Time
+	Attrs     LogAttributes
+	Partial   bool
+
+	// Err is an error associated with a message. Completeness of a message
+	// with Err is not expected, tho it may be partially complete (fields may
+	// be missing, gibberish, or nil)
+	Err error
+}
+
+// LogAttributes is used to hold the extra attributes available in the log message
+// Primarily used for converting the map type to string and sorting.
+type LogAttributes map[string]string
```

Comment:
If it's primarily used for converting to a slice and sorting that slice, could we make this a slice instead to avoid the transformation? If we need fast key lookups, or deduplication, it probably makes sense to keep it as a map, but I only see certain fixed keys like `<prefix>.node.id` being inserted. It seems like those could be appended to a slice instead, particularly if the next step is sorting.